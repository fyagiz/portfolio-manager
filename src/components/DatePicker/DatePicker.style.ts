import { StyleSheet } from "react-native";
import { COLOR, STYLE } from "../../utils/constants";

export default StyleSheet.create({
  iOScontainer: {
    flex: 1,
    justifyContent: "flex-end",
  },
  innerIOSContainer: {
    opacity: 1,
    backgroundColor: "black",
    borderRadius: STYLE.borderRadius,
  },
  iOSButtonStyle: {
    alignItems: "center",
    width: "100%",
    maxWidth: "auto",
  },
  textInputStyle: {
    color: COLOR.datePickerTextColor,
  },
});
