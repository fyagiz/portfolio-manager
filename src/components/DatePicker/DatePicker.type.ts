import { StyleProp, TextStyle } from "react-native";

export type DatePickerPropsType = {
  text: string;
  textInputStyle: StyleProp<TextStyle>;
  date: Date;
  onChange: (selectedDate: Date) => void;
  testID: string;
};
