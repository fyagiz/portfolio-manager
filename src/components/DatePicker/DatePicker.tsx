import { DatePickerPropsType } from "./DatePicker.type";
import { Modal, Platform, Pressable, Text, View, TextInput } from "react-native";
import { getBeforeDateByDay, getDateString, getToday } from "../../utils/helpers";
import DateTimePicker, { DateTimePickerEvent } from "@react-native-community/datetimepicker";
import { useState } from "react";
import Button from "../Button";
import styles from "./DatePicker.style";

const DatePicker = (props: DatePickerPropsType) => {
  const { text, date, onChange, textInputStyle, testID } = props;
  const [localDate, setLocalDate] = useState(date);
  const [isVisible, setIsVisible] = useState(false);
  const MAXIMUM_DATE = getToday();
  const MINIMUM_DATE = getBeforeDateByDay(MAXIMUM_DATE, 30);

  const onChangeAndroid = (event: DateTimePickerEvent, selectedDate?: Date) => {
    setIsVisible(false);
    onChange(selectedDate!);
  };

  const onChangeIOS = (event: DateTimePickerEvent, selectedDate?: Date) => {
    setLocalDate(selectedDate!);
  };

  const renderAndroidDatePicker = () => {
    return (
      isVisible && (
        <View>
          <View>
            <DateTimePicker
              testID={`${testID}DatePicker`}
              value={date}
              onChange={onChangeAndroid}
              mode="date"
              display="spinner"
              minimumDate={MINIMUM_DATE}
              maximumDate={MAXIMUM_DATE}
            />
          </View>
        </View>
      )
    );
  };

  const pressOutside = () => {
    setIsVisible(false);
  };

  const iOSConfirmButton = () => {
    onChange(localDate);
    setIsVisible(false);
  };

  const renderIOSDatePicker = () => {
    return (
      isVisible && (
        <Modal transparent animationType="slide">
          <Pressable style={styles.iOScontainer} onPress={pressOutside}>
            <View style={styles.innerIOSContainer}>
              <DateTimePicker
                testID={`${testID}DatePicker`}
                value={date}
                onChange={onChangeIOS}
                mode="date"
                display="spinner"
                minimumDate={MINIMUM_DATE}
                maximumDate={MAXIMUM_DATE}
              />
              <View>
                <Button testID={`${testID}DatePickerIOSButton`} text="Confirm" onPress={iOSConfirmButton} containerStyle={styles.iOSButtonStyle} />
              </View>
            </View>
          </Pressable>
        </Modal>
      )
    );
  };

  return (
    <View>
      <Text>{text}</Text>
      <Pressable
        onPress={() => {
          setIsVisible(true);
        }}
      >
        <View pointerEvents="none">
          <TextInput style={[styles.textInputStyle, textInputStyle]} editable={false} value={getDateString(date)} />
        </View>
      </Pressable>
      {Platform.OS === "android" ? renderAndroidDatePicker() : renderIOSDatePicker()}
    </View>
  );
};

export default DatePicker;
