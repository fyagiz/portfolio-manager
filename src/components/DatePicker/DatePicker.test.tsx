import { render } from "@testing-library/react-native";
import DatePicker from "./DatePicker";

/*
1- DatePicker should be rendered.
*/

describe("DatePicker Component Unit Tests", () => {
  test("should render correctly", () => {
    const today = new Date();
    const onChange = jest.fn();
    const datePickerComponent = render(<DatePicker date={today} onChange={onChange} text="Text" textInputStyle={{}} />);
    expect(datePickerComponent).toMatchSnapshot();
  });
});
