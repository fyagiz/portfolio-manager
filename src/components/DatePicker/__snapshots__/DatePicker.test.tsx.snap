// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker Component Unit Tests should render correctly 1`] = `
<View>
  <Text>
    Text
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      pointerEvents="none"
    >
      <TextInput
        editable={false}
        style={{}}
        value="15-02-2024"
      />
    </View>
  </View>
</View>
`;
